docs/source/configuration.rst

```diff
@@ -371,3 +371,16 @@ description
 -----------
 
 This argument describes the parameter.
+
+example
+-------
+
+This attribute specifies an optional example value for the parameter. If combined
+with 'choices', the example must be in the list of choices.
+
+schema
+------
+
+The schema argument is only supported by the BodyParameter. It specifies a
```

Comment:
please provide a use case and a detailed usage example

Reply:
Bodies can be complex data structures that can't be captured by a set of BodyParameters which address a single key-value pair. If you look at [seqrepo](https://glent-test-1.eu.seeds.basf.net/seqrepo-rest/services/documentation/swagger-ui.html#/), there is the /rest/v1/batchupload/ngs endpoint (upload-request-rest-controller) for which the body should be a list of objects (set of key-value pairs), which in turn have lists of objects for some of their values.

In order to support endpoints with such bodies in a manageable manner, they have been represented by a single BodyParameter. By adding a schema to this BodyParameter, qrest can validate whether the data structure that a user provides as value for this BodyParameter obeys the required data structure (similar to the validation that happens if the 'choices' attribute is used for query parameters). With the schema being attached to the BodyParameter, downstream applications can use it to generate templates that users can start from to build the data structure, or users can access the schema directly. 

For example,  ResourceConfig could look like this

```
class QrestExample(ResourceConfig):

    name = "qrest_example"
    path = ["qrest_example"]
    method = "POST"
    description = "qrest_example"

    schema = {
                "title": "create post",
                "description": "a schema to create a post",
                "type": "object",
                "properties": {
                    "user": {
                        "description": "id of the user",
                        "type": "string"
                    },
                    "body": {
                        "description": "the body of the post",
                        "type": "string"
                    }
                },
                "required": ["user", "body"]
     }

    example = {'user':'Alice', 'body':'Something about bob'}

    post = BodyParameter(name=None, required=True, example=example, schema=schema)
```

which leads to this at runtime

```
>>> edi.example_client.qrest_example.config.post.schema
{'title': 'create post', 'description': 'a schema to create a post', 'type': 'object', 'properties': {'user': {'description': 'id of the user', 'type': 'string'}, 'body': {'description': 'the body of the post', 'type': 'string'}}, 'required': ['user', 'body']}
>>> edi.seqrepo_client.qrest_example.config.post.example
{'user': 'Alice', 'body': 'Something about bob'}
>>> edi.seqrepo_client.qrest_example(post='foo')
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/dist-packages/qrest-3.1.2.dev0-py3.6.egg/qrest/resource.py", line 412, in check
  File "/usr/local/lib/python3.6/dist-packages/jsonschema/validators.py", line 353, in validate
    raise error
jsonschema.exceptions.ValidationError: 'foo' is not of type 'object'

Failed validating 'type' in schema:
    {'description': 'a schema to create a post',
     'properties': {'body': {'description': 'the body of the post',
                             'type': 'string'},
                    'user': {'description': 'id of the user',
                             'type': 'string'}},
     'required': ['user', 'body'],
     'title': 'create post',
     'type': 'object'}

On instance:
    'foo'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.6/dist-packages/qrest-3.1.2.dev0-py3.6.egg/qrest/resource.py", line 242, in __call__
  File "/usr/local/lib/python3.6/dist-packages/qrest-3.1.2.dev0-py3.6.egg/qrest/resource.py", line 253, in get_response
  File "/usr/local/lib/python3.6/dist-packages/qrest-3.1.2.dev0-py3.6.egg/qrest/resource.py", line 415, in check
qrest.exception.RestClientValidationError: value for post does not obey schema
```

The data structure in this example is still trivial ofcourse, and using a schema for it is probably overkill. But for structures like the ones in seqrepo, I think it is needed to make it easier on the enduser.

Reply:
I do understand the desire or need for a more thorough validation of the value of a body parameter or of the value of the body as a whole. 

Implementation-wise I would like to make the following remark. You add a `schema` parameter to the `BodyParameter` and unless that schema is `None`, you add behavior to a `BodyParameter`. Joris mentioned it already in an email: why not use a subclass of `BodyParameter`, e.g. `VerifiableBodyParameter`, that requires that parameter? That subclass can implement the additional functionality: you wouldn't have to change an existing class and the new class doesn't have to check whether the schema is `None`. It's almost as if this scenario is a poster-person for subclassing.